design-docs/commands-v3.md

```diff
@@ -0,0 +1,305 @@
+command.
+
+Each command run has its own coroutine to back its execution. When a command finishes, its backing
+continuation is done, and the command is removed from the scheduler. Any unfinished inner commands
+are cancelled at this point.
+
+### Factory Methods and Builders
+
+In contrast with v2’s decorator API that returns a new command object for each modification (which
+had some issues with naming and making telemetry difficult with deeply nested wrapper objects), v3
+uses builder objects that allow configuration to be set before creating a single command at the end.
+
+#### Forced Naming
+
+Builder objects will only permit command creation when a name for the command has been provided.
+
+#### Automatic Naming
+
+Parallel and sequential groups and builders permit automatically generated names to be used instead
+of manually specified ones.
+
+Automatic sequence names will simply be the list of all the commands in the sequence, separated by
+a “->” arrow sign; for example, “Step 1 -> Step 2 -> Step 3”
+
+Automatic parallel group names will be the list of the names of the commands in the group, separated
+by a pipe “|”. Required commands will be in their own group, separated from non-required commands.
+If either subgroup is empty, then there will be no indication in the output (ie, no empty group
+token like “()” will appear)
+
+### Cooperative Multitasking
+
+All commands *must* yield control back to the coroutine in their control loops. Because we cannot
+preempt a running command at any arbitrary point, the command must explicitly and deliberately cede
+control back to the scheduler in order for the framework to function. Commands may do this by
+calling `coroutine.yield()` on the injected coroutine object, which ultimately delegates to
+`Continuation.yield()` to pause the carrier continuation object.
+
+The scheduler has a single `ContinuationScope` object. When a command is mounted, the scheduler will
+create a new `Continuation` object bound to that scope, and a `Coroutine` object bound to the
+continuation. `ContinuationScope` and `Continuation` are currently JDK-private classes; the build
+needs to open the module to the wpilib/unnamed module for reflective access at compile and runtime.
+Because access is entirely reflective, wrapper classes at the library level will be used to make
+access easier.
+
+### Scheduling
+
+Scheduling a command with `Scheduler.schedule()` will add the command to a queue of pending
+commands. Any command in the queue that conflicts and has a lower priority will be removed from the
+queue without calling the `onCancel` hook. At the start of `Scheduler.run()`, all commands that
+conflict with those in the queue are cancelled, then the queued commands are promoted to running.
+
+Scheduling an inner command will bypass the queue and immediately begin to run. This avoids delays
+caused by loop timings for deeply nested commands.
```

Comment:
This also needs to specify behavior when scheduling conflicting children.

The current implementation throws an exception when scheduling a child that overlaps with an already running child command, which is different from top-level command scheduling (the running command would be interrupted and the new one would start up in the next cycle). That behavior was originally intended to prevent users from accidentally awaiting conflicting commands at the same time, but it's already handled by parallel and sequential groups using `Command.all`/`Command.parallel`/etc, and should be added to `Coroutine.awaitAll`/`awaitAny`.